### PR TITLE
Updates to allow bricsauthenticator to work in dev environment 

### PR DESCRIPTION
**Pull request by jcwomack**
These changes allow `BricsAuthenticator` to be used in a JupyterHub dev environment where authentication tokens are issued by a dev Keycloak instance (see [brics_jupyterhub_envs](https://github.com/isambard-sc/brics_jupyterhub_envs)).

The main change is to allow OpenID provider URL to be specified as a configurable trait for the `BricsAuthenticator` class.

The PR also adds code which attempts to JSON decode the `projects` claim after decoding by `jwt.decode()`. The `projects` claim value should itself be valid JSON, mapping project names to infrastructures. 

It seems the JSON in the claim is not reliably decoded (may come as a `dict` or `str`) by `jwt.decode()`. It is unclear whether this is an issue in PyJWT, or in how the claims are formed. 

The workaround code attempts to use `json.loads()` to convert the `projects` claim value to a Python object (`dict`). If the claim value is not something that can be decoded by `json.loads()` (e.g. already a `dict`), then the JSON decode is skipped and the claim is used as-is.

This is a workaround that may not be necessary in future, if the cause of the inconsistency when decoding the `projects` claim can be identified and rectified.

### CODE REVIEW

File: src/bricsauthenticator/bricsauthenticator.py [hunk 1]
```diff
@@ -21,11 +29,13 @@ async def get(self):
         # Log raw JWT token prior to decoding (optional for debugging)
         self.log.debug(f"Raw JWT Token: {token}")
 
-        oidc_server = "https://keycloak.isambard.ac.uk/realms/isambard"  # Hard-code to ensure it's a place we trust
-
         http_client = AsyncHTTPClient()
         try:
-            response = await http_client.fetch(request=f"{oidc_server}/.well-known/openid-configuration", method="GET")
+            # Log OIDC server request (optional for debugging)
+            self.log.debug(f"Requesting OIDC server configuration from {self.oidc_server}")
```
Comment (owhere) [inline src/bricsauthenticator/bricsauthenticator.py:35]: Even though this is a debug log, sensitive information might still be exposed if the logs are accessible to unauthorized users or if the logging level is set incorrectly in production.

Comment (jcwomack) [reply src/bricsauthenticator/bricsauthenticator.py:35]: This log message just shows the URL for the OIDC server (used to get the OIDC server configuration from the `.well-known/openid-config` endpoint. I believe this is a publicly accessible address. Is it sensitive information?

Comment (jcwomack) [reply src/bricsauthenticator/bricsauthenticator.py:35]: This log message just shows the URL for the OIDC server (used to get the OIDC server configuration from the `.well-known/openid-config` endpoint. I believe this is a publicly accessible address. Is it sensitive information?

Comment (jcwomack) [reply src/bricsauthenticator/bricsauthenticator.py:35]: OpenID configuration information endpoints for `keycloak.isambard.ac.uk` and `keycloak-dev.isambard.ac.uk` are publicly accessible:

* <https://keycloak.isambard.ac.uk/realms/isambard/.well-known/openid-configuration>
* <https://keycloak-dev.isambard.ac.uk/realms/isambard/.well-known/openid-configuration>

This repository is also public and open source, and the URL is easily findable in the source code: <https://github.com/isambard-sc/bricsauthenticator/blob/main/src/bricsauthenticator/bricsauthenticator.py>

**Review by jcwomack — COMMENTED** [review]
Ready to merge!

**Pull request merged**